### ISSUE TITLE
Pidgin freezes, telegram disconnects

### ISSUE THREAD

**Issue by ikuraj** [majn/telegram-purple#456]
Sporadically, pidgin freezes and fully locks one of my CPUs (100% utilization of one core). During this time, pidgin is effectively unresponsive, i.e. its interface is not rendered at all (keyboard shortcuts can be used). After some while, pidgin reports Telegram account has been disconnected. (Afterwards, Telegram account just reconnects after some time.)

Now, this is most probably an issue with telegram-purple, as pidgin is unfrozen when the account disconnects. Any idea for the cause of this? Also, how should I debug this issue, i.e. which messages should I look for in the log?

**Comment by BenWiederhake** [comment]
I also see the same issue, but never bothered to write it down in detail.  Thanks!

I have absolutely no idea what the cause is.  The only operation that should take significant CPU time is the "proof of work" step of login – but that happens immediately before asking for the SMS verification code, so it should be unrelated to this.

How to debug: no idea, otherwise I would have already started :P
I don't think that the log even continues during that time.  If you think otherwise (and that would be great!), start it like this `pidgin -d` to see the log even while everything is "stuck".  I hope this helps.

**Comment by ildar** [comment]
Care to run with gdb and get a back trace? Pidgin even has special docs on
it.

**Comment by ivankuraj** [comment]
Sure, I'll try to get the log and post it here asap. If that does not help, I'll try to recompile to allow interaction with gdb.

**Comment by ikuraj** [comment]
I think the log does not show anything of particular interests to this issue. The UI is effectively frozen while pidgin seemingly "struggles" to connect to the server, after which it just reports a connection failure. No particularly interesting logging messages related to telegram appearing in the process...

**Comment by rodneyrod** [comment]
Might have something to do with the backlog?

**Comment by BenWiederhake** [comment]
> while pidgin seemingly "struggles" to connect to the server, after which it just reports a connection failure.

Jup, then it's probably the "proof of work" part.  Sadly, libtgl uses a very inefficient way to do it, and in 32-bit it's even slower.  To make things worse, on Windows we're forced to 32-bit due to pidgin being 32-bit.

**Comment by rodneyrod** [comment]
@BenWiederhake Is it possible to increase the timeout for this on slower machines?

From what I can tell, there's a hard coded limit in the code for this, would it be possible to make that a variable that can be set in libpurple?

**Comment by BenWiederhake** [comment]
Absolutely, and in this context it sounds like a good idea.  Ideally, set the default higher on windows though `#if` magic.  Wanna write a PR? :)

**Comment by smnthermes** [comment]
Maybe this problem is caused by a high number of chats?

**Comment by smnthermes** [comment]
I replaced the previous account with another one that has few chats, and the problem went away.
